### PR TITLE
In which /vg/station13 gets high.

### PR DESCRIPTION
**Pull request by PJB3005**
We're rebasing to TG boys.

@MrStonedOne 

Current issues:
- [x] Initialization text doesn't have styling.
- [x] Something's up with lighting updates. They work but they're still strange as hell.
- [x] I need to test it more.
- [x] Crit seems to progress *much* faster. Double check this.
- [x] Emergency shuttle has lighting troubles.
- [x] Seems to be a runtime in faction setup.
- [x] RCDing a floor away doesn't make ZAS update. uh oh.
- [x] Oh yeah GC stats.

### CODE REVIEW

**Comment by BobdaBiscuit** [comment]
"IT JUST WORKS"

**Comment by SkowronX** [comment]
What does this mean

File: code/__HELPERS/unsorted.dm [hunk 1]
```diff
@@ -1584,4 +1584,22 @@ Game Mode config tags:
...
+/proc/stoplag()
+	. = round(1*DELTA_CALC)
+	sleep(world.tick_lag)
+	if (world.tick_usage > TICK_LIMIT_TO_RUN) //woke up, still not enough tick, sleep for more.
+		. += round(2*DELTA_CALC)
+		sleep(world.tick_lag*2*DELTA_CALC)
+		if (world.tick_usage > TICK_LIMIT_TO_RUN) //woke up, STILL not enough tick, sleep for more.
+			. += round(4*DELTA_CALC)
+			sleep(world.tick_lag*4*DELTA_CALC)
+			//you might be thinking of adding more steps to this, or making it use a loop and a counter var
```
Comment (ComicIronic) [inline code/__HELPERS/unsorted.dm]: Do it and make the number of attempts an argument for the proc, nerd boy.

**Comment by Probe1** [comment]
It means using a master controller that was written in the last year instead of five years ago and we will see some minor to moderate performance gains @SkowronX

**Comment by ComicIronic** [comment]
We're getting goofMC

File: code/__HELPERS/unsorted.dm (hunk 1 above)
Comment (PJB3005) [reply code/__HELPERS/unsorted.dm]: Just fuck my shit up.

**Comment by PJB3005** [comment]
StonedMC*

File: code/__HELPERS/unsorted.dm (hunk 1 above)
Comment (PJB3005) [reply code/__HELPERS/unsorted.dm]: Quick @MrStonedOne convince Comic to not enforce this.

File: code/controllers/mc/master.dm [hunk 2]
```diff
@@ -52,7 +52,6 @@ var/CURRENT_TICKLIMIT = TICK_LIMIT_RUNNING
 
 /datum/controller/master/New()
 	// Highlander-style: there can only be one! Kill off the old and replace it with the new.
-	check_for_cleanbot_bug()
```
Comment (9600bauds) [inline code/controllers/mc/master.dm]: Fantastic

File: vgstation13.dme [hunk 3]
```diff
@@ -272,7 +281,6 @@
 #include "code\game\shuttle_engines.dm"
 #include "code\game\skincmd.dm"
 #include "code\game\smoothwall.dm"
-#include "code\game\socket_talk.dm"
```
Comment (Kurfursten) [inline vgstation13.dme:275]: probably shouldn't disinclude this

File: vgstation13.dme [hunk 4]
```diff
@@ -1468,7 +1476,6 @@
 #include "code\modules\mob\living\silicon\mommi\soviet_mommi.dm"
 #include "code\modules\mob\living\silicon\mommi\update_icons.dm"
 #include "code\modules\mob\living\silicon\pai\death.dm"
-#include "code\modules\mob\living\silicon\pai\emote.dm"
```
Comment (Kurfursten) [inline vgstation13.dme:1472]: ditto

File: vgstation13.dme [hunk 5]
```diff
@@ -1528,7 +1535,6 @@
 #include "code\modules\mob\living\simple_animal\hostile\carp.dm"
 #include "code\modules\mob\living\simple_animal\hostile\crab.dm"
 #include "code\modules\mob\living\simple_animal\hostile\creature.dm"
-#include "code\modules\mob\living\simple_animal\hostile\decoy.dm"
```
Comment (Kurfursten) [inline vgstation13.dme:1532]: ditto

File: vgstation13.dme [hunk 6]
```diff
@@ -1551,7 +1557,6 @@
 #include "code\modules\mob\living\simple_animal\hostile\human\_humanoid.dm"
 #include "code\modules\mob\living\simple_animal\hostile\human\cult.dm"
 #include "code\modules\mob\living\simple_animal\hostile\human\diona.dm"
-#include "code\modules\mob\living\simple_animal\hostile\human\frostgolem.dm"
```
Comment (Kurfursten) [inline vgstation13.dme:1556]: ditto

File: vgstation13.dme (hunk 3 above)
Comment (PJB3005) [reply vgstation13.dme:275]: This is intended, no fucking clue about the other ones.

Probably left DM open to fix includes and then ruined it.

File: code/__HELPERS/unsorted.dm (hunk 1 above)
Comment (MrsTonedOne) [reply code/__HELPERS/unsorted.dm]: Number of attempts as an argument wouldn't make sense, waking up to just sleep again causes more lag.

Adding the while loop I once advised against when i first made this proc on the other hand wouldn't be that bad of an idea. its rare but some times when some shit spawns in a loop or shit that sleeps in a loop per item (throwing, and what not) byonds sleep queue can get so big that just the sleep operation to re-insert an item into the queue can have a high cost (5-10% of a 50ms tick per tick), so decreasing how often these wake up only to re-insert because the tick was already over by the time they woke up becomes worth it.

Most of this proc was written before stonedmc when tick_usage was new and scary.

Comment (MrsTonedOne) [reply code/__HELPERS/unsorted.dm]: The easiest way to see this in action that might still happen on /vg/ (depending) is dropping a really big bomb in mining. Throwing the sand and ore (assuming that happens on /vg/) can raise the sleep queue's count close to 100k.

(this is the kind of lesson you learn when you disable the bomb cap on the mining zlevel)

Comment (PJB3005) [reply code/__HELPERS/unsorted.dm]: We don't actually have explosions that throw stuff on here so I doubt that's applicable.

**Comment by MrsTonedOne** [comment]
\>net 1241 lines removed

**Comment by kol1th** [comment]
go back to tg

File: code/controllers/subsystem/air.dm [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+		air_master = new
+		air_master.Setup()
+
+	..()
+
+
+/datum/subsystem/air/fire(resumed = FALSE)
+	// No point doing MC_TICK_CHECK.
+	if (!air_processing_killed)
+		air_master.Tick()
```
Comment (MrsTonedOne) [inline code/controllers/subsystem/air.dm:35]: There would be if you inlined air_master.Tick() =P

Comment (PJB3005) [reply code/controllers/subsystem/air.dm:35]: that's for later.

**Comment by RemieRichards** [comment]
In the words of resident evil 7
"Welcome to the family son!"

**Comment by PJB3005** [comment]
I SAY SHIP IT!

**Comment by Probe1** [comment]
We'll merge it in 12 hours.  Everyone will be awake, including Pomf, in case anything goes wrong.

Do not merge this overnight when everyone is asleep.

Also god help our souls.

**Comment by MrsTonedOne** [comment]
Oh, pb&j

You need to make machine and obj's process() as well as mob's life() and other things be set waitfor = 0 so them sleeping doesn't hold up that entire subsystem and cause that one bug you experienced with sslighting.

(Assuming they aren't already....)

**Comment by PJB3005** [comment]
Oh shit good point.

**Comment by Probe1** [comment]
>its ready
 >its not ready
stop playing with my heart

**Comment by MrsTonedOne** [comment]
I mean, its a 2 line change.

set waitfor = 0 applies to child procs, so you can set it in the base process and life procs

**Comment by PJB3005** [comment]
@MrStonedOne that's implying we have a standardized process proc.

**Comment by MrsTonedOne** [comment]
add /datum/proc/process

regex /proc/process to /process

re-add /proc/ to /datum/process

**Comment by PJB3005** [comment]
@MrStonedOne not that simple.

**Comment by Sarcalogo** [comment]
Sarca finally won. /vg/ babies have abandoned Baybie code.

**Comment by Exxion** [comment]
>Falling for the meme description

File: code/__HELPERS/unsorted.dm (hunk 1 above)
Comment (MrsTonedOne) [reply code/__HELPERS/unsorted.dm]: tgstation/tgstation#22893 I caved myself.

Comment (PJB3005) [reply code/__HELPERS/unsorted.dm]: *ports*

**Comment by PKPenguin321** [comment]
Glad to see vg has finally been touched by the based god mso's light

Enjoy your no lag

**Comment by Probe1** [comment]
No lag is just a symptom of not enough FEATURES

**Comment by PJB3005** [comment]
Lag is actually the last reason I ported this.

**Comment by Kurfursten** [comment]
Please post a list of reasons in the OP of the PR in descending order of reasons you ported this

File: code/__HELPERS/unsorted.dm (hunk 1 above)
Comment (ComicIronic) [reply code/__HELPERS/unsorted.dm]: >do while

YECH

**Comment by optimumtact** [comment]
1) ongoing development effort on it
2) modern, proven to work with 100+ players
3) sort of well tested (who am I kidding right)

File: vgstation13.dme [hunk 8]
```diff
@@ -46,7 +50,6 @@
 #include "code\__HELPERS\iteration.dm"
 #include "code\__HELPERS\lists.dm"
 #include "code\__HELPERS\logging.dm"
-#include "code\__HELPERS\markov.dm"
```
Comment (PJB3005) [inline vgstation13.dme:49]: Fuck you too DM.

**Comment by PJB3005** [comment]
Should be ready for merge. I've created MR [`!1`](http://git.nexisonline.net/vgstation/secrets/merge_requests/1) on the secret repo to merge along with this so everything's updated correctly.

**Comment by Probe1** [comment]
We should push it to a test server.

File: __DEFINES/subsystem.dm [hunk 9]
```diff
@@ -1,21 +1,57 @@
...
+#define SS_PRIORITY_PIPENET        85
+#define SS_PRIORITY_LIGHTING       80
+#define SS_PRIORITY_AIR            70
+#define SS_PRIORITY_DISEASE        60
+#define SS_PRIORITY_FAST_MACHINERY 55
+#define SS_PRIORITY_EVENT          52
+#define SS_PRIORITY_UNSPECIFIED    50
+#define SS_PRIORITY_SUN            10
+#define SS_PRIORITY_GARBAGE        0
+#define SS_PRIORITY_INACTIVITY    -10
```
Comment (MrsTonedOne) [inline __DEFINES/subsystem.dm]: I don't know that the MC can handle negative/zero priorities properly, their primary use is weighing how the mc splits up a byond tick. if two (or more) subsystems need to run in one tick, it weights how much of the byond tick to give each subsystem based on this number, but its simple math.

I should maybe change the comment on that variable a touch to reflect that better.

Also i'm shocked a 0 priority subsystem isn't flooding you with divide by 0 runtimes. i'm 99% sure it divides by this number.

**Comment by d3athrow** [comment]
test server should be running it

File: __DEFINES/subsystem.dm (hunk 9 above)
Comment (PJB3005) [reply __DEFINES/subsystem.dm]: Oh I figured it worked in the same ordering way as init order and stuff.

Well then.

Oh also the display order is counter intuitive I'd say. Lower is higher.

Comment (MrsTonedOne) [reply __DEFINES/subsystem.dm]: Ya, that wasn't me, that was nullquery or neersighted i can't remember. They had intended it to be a ranking thing, 1 for first, 2 for second, etc.

I meant to fix it but that implies re-doing the orders on everything and ha! that sounds a lot like work.

**Comment by Megamaw** [comment]
what is this changing, in layman's terms? i want to make an educated decision before i decide that this is dumb, because right now it looks pretty dumb.

**Comment by BobdaBiscuit** [comment]
\>updating the mastercontroller to not be prehistoric is dumb

**Comment by Probe1** [comment]
Initial testing with <5 players on test server have shown no issues.

@Megamaw https://github.com/d3athrow/vgstation13/pull/13270#issuecomment-271695492

**Comment by Exxion** [comment]
@Megamaw In layman's terms, it is changing nothing.
There is nothing changing that is relevant to a layman.

Also, stop complaining about every change you don't understand

**Comment by MrsTonedOne** [comment]
@Megamaw It makes lighting update faster, other things update faster, and better balances what processes to run when during cpu lag spikes

**Comment by 9600bauds** [comment]
clip 2014-01-23 at 04 53 34

**Comment by SonixApache** [comment]
good job breaking code phrases nerd
